cluster/saltbase/salt/node-upgrade-gcp/node-upgrade-gcp.sh

```diff
@@ -0,0 +1,157 @@
+# is kicked off. Note that for any case other than performing an upgrade, it
+# returns, and thus must be called repeatedly to continue checking for upgrades.
+#
+# Assumed vars:
+#   WATCH_KEYS
+#   CACHE_DIR
+#   CACHE_EXT
+#   LATEST_EXT
+function monitor-upgradability() {
+  # Wait for _any_ metadata to change.
+  log "Waiting for metadata changes."
+  curl -s "http://metadata/computeMetadata/v1/instance/attributes/?recursive=true&wait_for_change=true" -H "Metadata-Flavor: Google"
+  result=$?
+  if [[ ${result} -ne 0 ]]; then
+    log "Metadata request failed; returned ${result}"
+    return
+  fi
+
+  # Get latest metadata.
+  log "Metadata changed; getting latest versions."
+  get-metadata "${LATEST_EXT}"
+
+  # Diff cache vs latest.
+  log "Diffing cached versus latest metadata."
+  local do_upgrade=0
+  for k in "${WATCH_KEYS[@]}"; do
+    diff "${CACHE_DIR}/${k}.${CACHE_EXT}" "${CACHE_DIR}/${k}.${LATEST_EXT}" > /dev/null
+    if [[ $? -ne 0 ]]; then
+      log "Metadata with key '${k}' differ."
+      do_upgrade=1
+      break
+    fi
+  done
+
+  # Don't upgrade if no metadata changed.
+  if [[ ${do_upgrade} -eq 0 ]]; then
+    log "No metadata differences found; will not upgrade."
+    return
+  fi
+
+  # Kick off a node upgrade.
+  log "Relevant metadata differ; performing node upgrade."
+  upgrade-node
+}
+
+# upgrade-node launches a node upgrade, assuming that the script to do so is
+# found in ${CACHE_DIR}/startup-script.${LATEST_EXT}, then exits.
+#
+# Assumed vars:
+#   CACHE_DIR
+#   LATEST_EXT
+function upgrade-node() {
+  log "Running ${CACHE_DIR}/startup-script.${LATEST_EXT}"
+  # Wait 1 second (generous) for this upgrader to exit.
+  sudo /bin/bash -c "sleep 1; /bin/bash '${CACHE_DIR}/startup-script.${LATEST_EXT}' --push" &
+  # The upgrade must start this process again.
+  log "Exiting"
+  exit 0
+}
+
+# Execution begins here.
+
+# This is running in a root-owned directory, so we do two sudo calls to create
+# a nonprivileged one.
+log "Creating cache directory '${CACHE_DIR}/'"
+sudo mkdir -p "${CACHE_DIR}"
+sudo chown $(whoami) "${CACHE_DIR}"
+
+# If we just did an upgrade, use the upgraded-to values as the current cache.
+# Otherwise, grab the current metadata as the cache (before the first upgrade,
+# any metadata changes that happens before this runs won't be caught).
+log "Checking for cached 'latest' configuration metadata."
+latest_exists=1
+for k in "${WATCH_KEYS[@]}"; do
+  if [[ ! -s "${CACHE_DIR}/${k}.${LATEST_EXT}" ]]; then
+    latest_exists=0
+  fi
+done
+if [[ latest_exists -eq 1 ]]; then
+  log "Found 'latest' versions of all required metadata. Using as current config."
+  for k in "${WATCH_KEYS[@]}"; do
+    mv "${CACHE_DIR}/${k}.${LATEST_EXT}" "${CACHE_DIR}/${k}.${CACHE_EXT}"
+  done
+else
+  log "One or more 'latest' metadata files not found locally. Getting from server."
+  get-metadata "${CACHE_EXT}"
+fi
+
+# Wait indefinitely for upgrades.
+while true; do
```

Comment:
Bash is awesome as long as you need 20-30 lines of it, never need arrays or structs, and don't mid debugging it each time you use it.